pkg/storage/etcd3/etcdwatcher/watcher.go

```diff
@@ -0,0 +1,340 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcdwatcher
+
+import (
+	"container/list"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/storage"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/coreos/etcd/clientv3"
+	etcdrpc "github.com/coreos/etcd/etcdserver/api/v3rpc/rpctypes"
+	"golang.org/x/net/context"
+)
+
+type Watcher interface {
+	// Watch watches on a key and returns a watch.Interface that transfer relevant notifications.
+	// If rev is zero, it will try to get existing objects if any and then watch after that revision.
+	// If rev is non-zero, it will watch from given revision.
+	// If recursive is false, it watches on given key.
+	// If recursive is true, it watches any children and directories under the key, excluding the root key itself.
+	// filter must be non-nil. Only if filter returns true will the changes be returned.
+	Watch(ctx context.Context, key string, rev int64, recursive bool, filter storage.FilterFunc) (watch.Interface, error)
+}
+
+// New returns a Watcher.
+func New(client *clientv3.Client, codec runtime.Codec, versioner storage.Versioner) Watcher {
+	return newWatcher(client, codec, versioner)
+}
+
+// watcher implements Watcher.
+type watcher struct {
+	client    *clientv3.Client
+	codec     runtime.Codec
+	versioner storage.Versioner
+}
+
+// watchChan implements watch.Interface.
+type watchChan struct {
+	watcher       *watcher
+	key           string
+	initialRev    int64
+	recursive     bool
+	filter        storage.FilterFunc
+	ctx           context.Context
+	cancel        context.CancelFunc
+	resultChan    chan watch.Event
+	eventChan     chan *event
```

Comment:
IIUC, this is incoming stream of events from etcd, right? How about calling this "incomingEventChan"?